nixos/modules/config/nix.nix

```
@@ -386,5 +387,12 @@ in
         )
       );
     };
+
+    # Enable git if flake support is enabled
+    programs.git = mkIf (builtins.match ".*flakes.*" cfg.extraOptions != null) ({
```

Please prefer matching via `nixpkgs.flake.source` being set or `nix.settings.experimental_features` containing "flakes". This current setup isn't likely to work on all cases. Honestly my preference would be `nixpkgs.flake.source` since that _will_ be set in a flake exposed system configuration.

---

Not necessarily, if the system configuration isn't built using flakes but still enables the experimental feature, `nixpkgs.flake.source` won't be defined.

Additionally, this is the second time I've heard about `nix.settings.experimental_features`. I can't find any reference to it (`git grep experimental_features` gives nothing). Mind linking to the docs?

---

`nix.settings` directly maps to `/etc/nix/nix.conf` settings. Those settings can be viewed at `man 5 nix.conf`, which contains an entry about `experimental_features`.

Also that's a fair point, prefer checking if `nix.settings.experimental_features` contains "flakes" in its list.

---

```suggestion
    programs.git = mkIf (lib.elem "flakes" (cfg.settings.experimental-features ++ cfg.settings.extra-experimental-features)) ({
``` don't forget `extra-`

---

`nix.settings` is a free‐form option, which means it accepts arbitrary Nix options that may not necessarily be defined in NixOS.

---

FWIW I've only ever seen flakes being enabled using `nix.extraOptions = "experimental-features = nix-command flakes";` (including in tutorials and my own system flake). I'll definitely change that to `nix.settings.experimental-features` as that's a lot cleaner, and I'll also implement this change, but I'll or it with the original logic to catch flakes being enabled that way.

---

We can’t process people’s configuration properly if they opt out of the structured configuration mechanism. Running regexes on generated text configuration files worries me, especially for a minor convenience like this.

---

Fair enough, although I am curious why running regexes worries you.

> We can’t process people’s configuration properly if they opt out of the structured configuration mechanism. 

If you want, I can make a proper regex for this, and then we *can*.

---

While that might work for `nix.conf`, I’m more worried about the bottom of the slippery slope where we’re trying to parse XML in Nix to handle niceties for people who explicitly opted out of the configuration system…

The whole point of NixOS having structured configuration is that if people buy into it their configuration can be processed in a structured way.

---

In general, I agree. If it'll make it easier to merge this, I'll drop that code.